code/datums/quirks/_quirk.dm

```diff
@@ -13,27 +13,87 @@
 	///Amount of points this trait is worth towards the hardcore character mode; minus points implies a positive quirk, positive means its hard. This is used to pick the quirks assigned to a hardcore character. 0 means its not available to hardcore draws.
 	var/hardcore_value = 0
 	var/mob/living/quirk_holder
+	/// This quirk should START_PROCESSING when added and STOP_PROCESSING when removed.
+	var/processing_quirk = FALSE
+
+/datum/quirk/Destroy()
+	if(quirk_holder)
+		remove_from_current_holder()
+
+	return ..()
+
+/// Called when quirk_holder is qdeleting. Simply qdels this datum and lets Destroy() handle the rest.
+/datum/quirk/proc/on_holder_qdeleting(mob/living/source, force)
+	SIGNAL_HANDLER
+	qdel(src)
+
+/**
+ * Adds the quirk to a new quirk_holder.
+ *
+ * Performs logic to make sure new_holder is a valid holder of this quirk.
+ * Returns FALSEy if there was some kind of error. Returns TRUE otherwise.
+ * Arguments:
+ * * new_holder - The mob to add this quirk to.
+ * * quirk_transfer - If this is being added to the holder as part of a quirk transfer. Quirks can use this to decide not to spawn new items or apply any other one-time effects.
+ */
+/datum/quirk/proc/add_to_holder(mob/living/new_holder, quirk_transfer = FALSE)
+	if(!new_holder)
+		CRASH("Quirk attempted to be added to null mob.")
+
+	if(human_only && !ishuman(new_holder))
+		CRASH("Human only quirk attempted to be added to non-human mob.")
+
+	if(new_holder.has_quirk(type))
+		CRASH("Quirk attempted to be added to mob which already had this quirk.")
+
+	if(quirk_holder)
+		CRASH("Attempted to add quirk to a holder when it already has a holder.")
+
+	quirk_holder = new_holder
+	quirk_holder.quirks += src
 
-/datum/quirk/New(mob/living/quirk_mob, spawn_effects)
-	..()
-	if(!quirk_mob || (human_only && !ishuman(quirk_mob)) || quirk_mob.has_quirk(type))
-		qdel(src)
-		return
-	quirk_holder = quirk_mob
-	SSquirks.quirk_objects += src
-	to_chat(quirk_holder, gain_text)
-	quirk_holder.roundstart_quirks += src
 	if(mob_trait)
-		ADD_TRAIT(quirk_holder, mob_trait, ROUNDSTART_TRAIT)
-	START_PROCESSING(SSquirks, src)
+		ADD_TRAIT(quirk_holder, mob_trait, QUIRK_TRAIT)
+
 	add()
-	if(spawn_effects)
-		on_spawn()
-	if(quirk_holder.client)
-		post_add()
-	else
-		RegisterSignal(quirk_holder, COMSIG_MOB_LOGIN, .proc/on_quirk_holder_first_login)
 
+	if(processing_quirk)
+		START_PROCESSING(SSquirks, src)
```

Comment:
It might just be better letting quirks handle their own processing, giving them more fine control over when they process and when they don't.

Reply:
See other comment on micro-optimising quirk processing()